### DISCUSSION TITLE
Classes assigned seemingly on their own.

### DISCUSSION THREAD

**Discussion by giacarrea** [AvaloniaUI/Avalonia#13908] [Q&A] [answered]
Hello, I'm busy trying to port an existing app from avalonia 10 to 11 and I'm having the following issue:

I have a class specific styling for button  mixed in with other controls, but when switching to avalonia 11,  some buttons that weren't assigned to that class in my axaml file were somehow assigned to it regardless - as I could notice in the ui dev tool.

Weirder still, when I split my styling over two différent styles, it then worked as intended... (see screenshot)
![image](https://github.com/AvaloniaUI/Avalonia/assets/60611503/6eeafddb-2a30-44f6-adba-3b2ef000e3a7)

Also, the same window presenting the issue will be displayed correctly the first time its rendered. If I close it, open another one, then reopen the first one, then I get my styling issue...

Similarly, I have a button with a flyout menu; if I remove the menuItem, I don't have the styling issue anymore (which happens on another window...)

![image](https://github.com/AvaloniaUI/Avalonia/assets/60611503/93c6e4e3-08fc-44b1-979a-c5eb6005ba70)

Has anyone ever had a similar issue or know where it could come from?

**Comment by giacarrea** [comment]
Here's an app - derived from the avalonia sample project - reproducing the issue:
[Avalonia.Samples-main.zip](https://github.com/AvaloniaUI/Avalonia/files/13636880/Avalonia.Samples-main.zip)

Start the BasicMvvmSample project, then in the "PlotView" tab, click the admin menu, then the "Control panel" button. The first two rows of button in the displayed window will have the "SingleLineButton" style assigned even tough they shouldn't. (If this isn't the case, try closing the window, then re-opening it)


![image](https://github.com/AvaloniaUI/Avalonia/assets/60611503/f98a8639-9fee-489b-8f87-42f50f3823ae)

**Reply by timunie** [reply]
Please check style docs and trouble shooting page. If it still doesn't help you will have to extract a very minimum sample to reproduce.

**Reply by maxkatz6** [reply, marked as answer]
It is a bug caused by "MenuItem" being part of the selector.
I.e. having style like "Button.someClass, MenuItem" - Button will be matched as "sometimes" (and thus - it will be re-evaluated to make sure it actually has a class), and MenuItem will be matched as "always" - it won't be re-evaluated anymore in runtime (unless style is removed later). It would work, but style results caching break this, as it mistakenly uses "always" result for a button as well.

**Reply by timunie** [reply]
Ah okay 👍

**Reply by giacarrea** [reply]
So, just isolating my MenuItem styles shouldn't cause issues in the meantime?

**Reply by timunie** [reply]
Exactly, yes.
